packages/mui-material/src/Switch/Switch.js

```diff
@@ -140,33 +148,38 @@ const SwitchSwitchBase = styled(SwitchBase, {
         backgroundColor: 'transparent',
       },
     },
-    ...(ownerState.color !== 'default' && {
-      [`&.${switchClasses.checked}`]: {
-        color: (theme.vars || theme).palette[ownerState.color].main,
-        '&:hover': {
-          backgroundColor: theme.vars
-            ? `rgba(${theme.vars.palette[ownerState.color].mainChannel} / ${
-                theme.vars.palette.action.hoverOpacity
-              })`
-            : alpha(theme.palette[ownerState.color].main, theme.palette.action.hoverOpacity),
-          '@media (hover: none)': {
-            backgroundColor: 'transparent',
+    variants: [
+      ...Object.entries(theme.palette)
+        .filter(([, value]) => value.main && value.light) // check all the used fields in the style below
+        .map(([color]) => ({
+          props: { color },
+          style: {
+            [`&.${switchClasses.checked}`]: {
+              color: (theme.vars || theme).palette[color].main,
+              '&:hover': {
+                backgroundColor: theme.vars
+                  ? `rgba(${theme.vars.palette[color].mainChannel} / ${theme.vars.palette.action.hoverOpacity})`
+                  : alpha(theme.palette[color].main, theme.palette.action.hoverOpacity),
+                '@media (hover: none)': {
+                  backgroundColor: 'transparent',
+                },
+              },
+            },
+            [`&.${switchClasses.disabled}`]: {
```

Comment:
```suggestion
              [`&.${switchClasses.disabled}`]: {
            },
```
`disabled` style should be in `checked` to fix this:

<img width="1064" alt="image" src="https://github.com/mui/material-ui/assets/18292247/a112f8cb-3eb8-41b4-b5f3-8c3fb0671599">